Allow outgoing webhook to add reaction when sending a response without additional calls to the REST API.

PIG208 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
Fixes: #12811

**Testing plan:** <!-- How have you tested? -->
It has been tested both manually and with automated tests.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->
This introduces a new format for response messages for outgoing webhooks:
```json
{
  "reactions": [
    {"emoji_name": "wave"}
  ]
}
```
It also works with `content`:
```json
{
  "content": "test",
  "reactions": [
    {"emoji_name": "wave"}
  ]
}
```
Responding with two emoji reactions along with a message:
![image](https://user-images.githubusercontent.com/39874143/114476115-3e246a00-9c2c-11eb-83db-0da7ffd8b0b5.png)
When the `emoji_name` field is missing in a specified emoji reaction:
![image](https://user-images.githubusercontent.com/39874143/114476076-2b119a00-9c2c-11eb-8a4e-8ae978686225.png)
Updated documentation for the API:
![image](https://user-images.githubusercontent.com/39874143/114475940-e38b0e00-9c2b-11eb-9df2-0bec154e4aa2.png)
Error messages;
![image](https://user-images.githubusercontent.com/39874143/114477444-be4bcf00-9c2e-11eb-92d4-d700662e3c02.png)
<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

PIG208 committed:
outgoing webooks: Support adding reactions in response.

send_response_reaction is created for the sake of this feature, which
is implemented in a way similar to `send_response_message`. Since this
change, the outgoing webhooks will be able to respond to messages only
with reactions, and sending a response message at the same time is
also supported.

Fixes: #12811

PIG208 commented:
While working on the PR, I found that if the response message is not JSON, it will send an error message after the message that triggers it, but it doesn't notify the user. This might be inconsistent with our error handling logic for outgoing webhooks. Prove me wrong if it's intended this way.
```python
try:
    response_json = json.loads(response.text)
except json.JSONDecodeError:
    fail_with_message(event, "Invalid JSON in response")
    return
```

timabbott commented:
Good find; I think we ended up doing just that in b7b1ec0aebbf75c2f66a4c81a42bf4da0ac7c9e1.

timabbott commented:
A small comment on commit messages: You should only put `Fixes: ` in one commit.  Also, documentation changes generally belong in the commit that changes reality to match that documentation; so we should squash the documentation commit into the previous one.

PIG208 commented:
Thank you for the feedback!

timabbott commented:
I merged the first commit, after moving `check_add_reaction` after `do_add_reaction`, since generally it's clearer to have functions call functions above them in the file.

I pushed the second commit back here after a few small documentation/comment changes.  It looks correct, but I'd like tests that actually verify the full set of parameters passed to `check_add_reaction` -- one can imagine some future refactoring bug where we mess up the plumbing of parameters, or something, and so it's worth having a test.  (Probably in `test_outgoing_webhook_system.py`?)

PIG208 commented:
It looks like that I have broken something with the rebase.

PIG208 committed:
messages: Allow service bots to access private messages.

This commit allows service bots including embedded bots and outgoing
webhook bots to access the private messages they sent or received in
a group or a 1-to-1 chat.

Related: #13658

PIG208 commented:
I've rebased this due to a recent change. I'm not sure if a service bot is the only case when a user doesn't have a UserMessage row. It is possible that we can remove the check below?
```python
if (
   user_profile.bot_type == UserProfile.EMBEDDED_BOT
   or user_profile.bot_type == UserProfile.OUTGOING_WEBHOOK_BOT
):
```

eeshangarg commented on zerver/lib/message.py:
```diff
@@ -696,6 +698,22 @@ def has_message_access(
+        if (
+            user_profile.bot_type == UserProfile.EMBEDDED_BOT
+            or user_profile.bot_type == UserProfile.OUTGOING_WEBHOOK_BOT
+        ):
```
This can just be an `elif` statement and be much clearer.

eeshangarg commented on zerver/lib/outgoing_webhook.py:
```diff
@@ -205,6 +211,39 @@ def send_response_message(
+    bot_id is the user_id of the bot sending the response
+
+    message_info is used to address the message to be added with the reaction and
+    should have at least these fields:
```
Better to phrase this as "message_info contains information that identifies the message to add the reaction to. This parameter should have at least the following fields:"

eeshangarg replied:
Or is this the message that one adds in reply along with the reactions to the previous message?

PIG208 replied:
Good advice, there is indeed some confusion here. Just changed the docstring.

eeshangarg reviewed (commented):
Hey @PIG208! This looks great, thanks for working on this! I just left a couple of minor comments.
As always, please feel free to reach out if you have any questions. Thanks! :)

eeshangarg commented:
@PIG208 I wonder why the build is failing here. Could you please investigate?

PIG208 commented:
> @PIG208 I wonder why the build is failing here. Could you please investigate?

I have rerun the tests several times and I think this might be a nondeterministic error? Will rerun it again.

timabbott commented:
Does https://github.com/zulip/zulip/pull/32247 mean we should close this? I can't remember if there was some other detail we need to manage in order to allow adding reactions.

PIG208 commented:
I think the difference here is that #12811 asks for an API change that supports adding a reaction through a response, not with a separate request. If we are to implement this, though, we might just start from scratch and rewrite most of this.
